problem fixes

@claude

please fix problems in the code following these instructions:

0) Run order (so the cascade of errors collapses fast)

Module resolution & path aliases (fixes hundreds of “Cannot find module …” + “need explicit file extensions”): §1.

TypeScript compiler options alignment (verbatimModuleSyntax, exactOptionalPropertyTypes, moduleResolution): §2.

Undefined value narrowing & runtime validation (cookies, querystrings, envs, CSV utils): §3.

Implicit any & incorrect component typings (icons, reducers, map callbacks): §4.

API/web shared types & barrels: §5.

Library/dep gaps & stray ts-expect-error: §6.

1) Module resolution & path aliases
1A. “Cannot find module '@/…'” across web & api

Create/normalize path aliases and ensure build tools know them.

Actions (AI-safe edits):

Add baseUrl and paths in the repo-root tsconfig.base.json (or create it) and reference from app tsconfigs.

{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@/*": ["apps/web/src/*"],
      "@api/*": ["apps/api/src/*"]
    }
  }
}


In Vite (web): ensure alias matches TS:

// apps/web/vite.config.ts
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

export default defineConfig({
  plugins: [react()],
  resolve: {
    alias: { '@': '/apps/web/src' }
  }
})


In ts-node/tsc for API: either import via @api/... or keep relative paths. If keeping @api, add a module resolver (tsconfig-paths) in dev, or stick to relative imports.

Errors this resolves: missing @/... modules in App.tsx, admin/shop components, lib utils/types, etc.

1B. Node16/NodeNext requiring explicit .js extensions in relative ESM imports

Two safe options; choose exactly one:

Option 1 (recommended): Switch web to moduleResolution: "bundler" so you don’t add .js to TS paths, which plays nicer with Vite.

Option 2: Keep node16 and add .js extensions to relative imports in source (TS will rewrite on emit if you also set allowImportingTsExtensions: true—not ideal).

AI step (Option 1): in apps/web/tsconfig.json:

{
  "compilerOptions": {
    "moduleResolution": "bundler",
    "verbatimModuleSyntax": true
  }
}


If you prefer Option 2, auto-edit all flagged imports (seen in many index.ts, feature barrels, and hooks). Examples flagged: CardsTab.tsx line 18 and many others.

2) TS compiler options alignment

You’re on verbatimModuleSyntax and exactOptionalPropertyTypes, which expose stricter typing (good), but several files aren’t written for them.

Actions:

Keep both options on (preferred), then fix call sites & types as below.

Alternatively, relax exactOptionalPropertyTypes to reduce noise, but we’ll fix instead.

Problems caused by these flags appear in logger.ts (pino transport), client.ts (params/schema optional), OrdersTab.tsx lucide icon prop typing, etc.

3) Undefined value narrowing & runtime validation
3A. API: cookies, headers, env, and Buffer.from

Many calls pass string | undefined into strict signatures (e.g., Buffer.from, zod schemas, SQL params).

Actions (mechanical):

Guard and narrow before use; never non-null assert.

Examples to apply:

// apps/api/src/middleware/auth.ts
const raw = req.headers['cookie'];
if (typeof raw !== 'string' || !raw.length) {
  return res.status(401).json({ error: 'Missing cookie' });
}
const buf = Buffer.from(raw, 'utf8'); // now string


Fixes multiple lines in auth.ts.

For URLSearchParams/req.query in inventory routes:

// apps/api/src/routes/inventory.ts
const pageRaw = req.query.page;
const page = typeof pageRaw === 'string' ? Number(pageRaw) : 1;
if (Number.isNaN(page)) return res.status(400).json({ error: 'Invalid page' });


Also guard strings used as IDs (e.g., id ?? return 400) before DB calls. Addresses multiple “string | ParsedQs” and “string | undefined” errors.

3B. Object may be undefined before Object.keys / destructures

Guard first:

const row = result.rows[0];
if (!row) return res.status(404).json({ error: 'Not found' });
const keys = Object.keys(row);


Fixes Object.keys overload error.

3C. Web CSV utils & ValidationResult shape

Ensure you don’t call Object.keys(undefined).

Make ValidationResult return the full shape (totalRows, validRows, booleans).

type ValidationResult = {
  totalRows: number;
  validRows: number;
  valid: boolean;
  errors: string[];
  warnings: string[];
};

// When failing early, still return full shape:
return { totalRows: 0, validRows: 0, valid: false, errors, warnings };


Also, treat CSV field access as optional:

const header = record?.header ?? '';
if (!header) errors.push('Missing header');


Resolves many csv.ts errors.

3D. Error handling with unknown

Where catch (error: unknown):

if (error && typeof error === 'object') {
  const { message, stack } = error as { message?: string; stack?: string };
  log({ name: 'Error', message: message ?? 'Unknown', stack });
}


Fixes error is unknown occurrences in error handler and migration scripts.

4) Implicit any & incorrect component typings
4A. Add param types for reducers/maps

Add explicit types to callback params (acc, item, card, etc.) in AnalyticsTab.tsx, ShopCart.tsx, RecentlyViewedCards.tsx, ShopPage*.tsx:

items.reduce((acc: number, item: CartItem) => acc + item.qty, 0);


4B. Lucide icons Type mismatch

ComponentType<{ className?: string }> is too narrow for lucide’s ForwardRefExoticComponent<SVGProps<…>>.

Actions (either):

Type as React.ComponentType<React.SVGProps<SVGSVGElement>>, or

Wrap icons with a tiny adapter:

type IconLike = React.ComponentType<React.SVGProps<SVGSVGElement>>
const columns: Array<{icon: IconLike; ...}> = [...]


Fixes multiple OrdersTab.tsx icon assignments.

5) Shared types, barrels, and type-only imports
5A. Add missing shared types exports

A lot of files import @/types and specific model types. Ensure apps/web/src/types/index.ts re-exports concrete values or namespaces where needed, and use type-only imports with verbatimModuleSyntax.

Fix type-only imports:

// before
import { ReactNode, ErrorInfo } from 'react';
// after
import type { ReactNode, ErrorInfo } from 'react';


Targets: ErrorBoundary.tsx, Toast.tsx, and others.

5B. “export default” of a type only

types/models/card.ts: export default cannot reference a type with verbatimModuleSyntax.

Action: remove default export for types, use named export type Card = …, and update imports.

Automated edits the AI system can apply
A) Add/normalize tsconfig files

tsconfig.base.json (root) as in §1A.

apps/web/tsconfig.json ensure:

{
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "bundler",
    "verbatimModuleSyntax": true,
    "exactOptionalPropertyTypes": true
  },
  "extends": "../../tsconfig.base.json"
}


apps/api/tsconfig.json ensure:

{
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "node16",
    "verbatimModuleSyntax": true,
    "exactOptionalPropertyTypes": true,
    "paths": { "@api/*": ["./src/*"] }
  },
  "extends": "../../tsconfig.base.json"
}

B) Codemod: add type-only imports for React types

Pattern: replace import { ReactNode|ErrorInfo|FC } from 'react' with import type { … } from 'react'.

Targets include: ErrorBoundary.tsx, Toast.tsx.

C) Codemod: guard Object.keys(x) and .from(x)

Find Object.keys( where arg may be union with undefined; wrap with if (!x) ….

Replace Buffer.from(maybeString) with:

if (typeof maybeString !== 'string') { /* handle */ }
Buffer.from(maybeString, 'utf8');


Targets: inventory.ts, csv.ts, tests using cookies.

D) Codemod: icon prop typing

Search in OrdersTab.tsx for ComponentType<{ className?: string }> and replace with React.ComponentType<React.SVGProps<SVGSVGElement>>.

E) Web relative import extensions (only if you stay on node16)

Append .js in relative imports flagged by the log (e.g., ../hooks/useCart → ../hooks/useCart.js). If you adopt moduleResolution: "bundler" (recommended), skip this step. Examples: features/hooks/index.ts, many shared/*/index.ts.

F) CSV ValidationResult return shape & field guards

Ensure all return paths populate { totalRows, validRows, valid, errors, warnings }.

Guard optional fields before string ops.

G) Tests: cookie header type narrowing

In apps/api/tests/*checkout*.test.ts and integration/auth.routes.test.ts:

When extracting cookie from response, assert presence and branch:

const cookie = res.headers['set-cookie'];
expect(Array.isArray(cookie) && cookie[0]).toBeTruthy();
return request(app).get('/route').set('Cookie', cookie as string[]);


Fixes multiple overload errors on .set('Cookie', ...).

Spot fixes (representative snippets)

apps/web/src/lib/api/client.ts (exactOptionalPropertyTypes):
Replace objects like { params: maybeParams, schema: maybeSchema } with conditional spreads:

const opt: {
  params?: Record<string, unknown>
  body?: unknown
  headers?: HeadersInit
  schema?: ZodType<T>
} = {
  ...(params ? { params } : {}),
  ...(schema ? { schema } : {})
}


Resolves TS2379 across several lines.

apps/web/src/shared/ui/CurrencySelector.tsx: “contains on type never”
Ensure selectedCurrency is typed and not narrowed to never. E.g.:

type Currency = 'NZD'|'USD'|'AUD'
const selectedCurrency: Currency | null = props.value ?? null;
if (selectedCurrency && list.includes(selectedCurrency)) { /* … */ }


Fixes .contains on never.

apps/web/src/shared/media/OptimizedImage.tsx: 'entry' is possibly undefined
Guard IntersectionObserver entries:

const observer = new IntersectionObserver(([entry]) => {
  if (!entry) return;
  if (entry.isIntersecting) setVisible(true);
});

--- a/apps/api/src/middleware/auth.ts
+++ b/apps/api/src/middleware/auth.ts
@@ -108,6 +108,13 @@ export const adminAuthBasic = async (
     }
 
     const base64Credentials = authHeader.split(" ")[1];
+    if (typeof base64Credentials !== 'string' || !base64Credentials.length) {
+      res.status(401).json({
+        error: "Invalid authorization header format",
+        hint: "Use Basic Auth with username:password"
+      });
+      return;
+    }
     const credentials = Buffer.from(base64Credentials, "base64").toString("utf-8");
     const [username, password] = credentials.split(":");
 

--- a/apps/api/src/routes/inventory.ts
+++ b/apps/api/src/routes/inventory.ts
@@ -12,10 +12,16 @@ const router = express.Router();
  */
 router.get("/admin/inventory", adminAuthJWT, async (req: Request, res: Response) => {
   try {
-    const { card_id, page = "1", per_page = "50" } = req.query;
-    
-    const limit = Math.min(parseInt(per_page as string, 10) || 50, 100);
-    const offset = (parseInt(page as string, 10) - 1) * limit;
+    const { card_id, page, per_page } = req.query;
+
+    const pageRaw = page;
+    const pageNum = typeof pageRaw === 'string' ? Number(pageRaw) : 1;
+    const currentPage = Number.isNaN(pageNum) || pageNum < 1 ? 1 : pageNum;
+
+    const perPageRaw = per_page;
+    const perPageNum = typeof perPageRaw === 'string' ? Number(perPageRaw) : 50;
+    const limit = Math.min(Number.isNaN(perPageNum) ? 50 : perPageNum, 100);
+    const offset = (currentPage - 1) * limit;
 
     let sql = `
       SELECT 
@@ -56,7 +62,7 @@ router.get("/admin/inventory", adminAuthJWT, async (req: Request, res: Response)
     return res.json({
       inventory,
       pagination: {
-        page: parseInt(page as string, 10),
+        page: currentPage,
         per_page: limit,
         total,
         total_pages: Math.ceil(total / limit),
@@ -295,7 +301,11 @@ router.get("/admin/inventory/export", adminAuthJWT, async (req: Request, res: Re
       return res.status(404).json({ error: "No inventory to export" });
     }
 
-    const headers = Object.keys(inventory[0]).join(",");
+    const row = inventory[0];
+    if (!row) {
+      return res.status(500).json({ error: "Failed to process inventory data" });
+    }
+    const headers = Object.keys(row).join(",");
     const rows = inventory.map(row => 
       Object.values(row).map(val => 
         typeof val === "string" && val.includes(",") ? `"${val}"` : val

--- a/apps/api/tests/integration/auth.routes.test.ts
+++ b/apps/api/tests/integration/auth.routes.test.ts
@@ -52,8 +52,9 @@ describe("POST /api/auth/admin/login", () => {
       .expect(200);
 
     expect(res.body.success).toBe(true);
-    expect(res.headers["set-cookie"]).toBeDefined();
-    expect(res.headers["set-cookie"][0]).toContain("adminToken");
+    const cookies = res.headers["set-cookie"];
+    expect(Array.isArray(cookies) && cookies[0]).toBeTruthy();
+    expect(cookies![0]).toContain("adminToken");
   });
 });
 
@@ -64,8 +65,9 @@ describe("POST /api/auth/admin/logout", () => {
       .expect(200);
 
     expect(res.body.success).toBe(true);
-    expect(res.headers["set-cookie"]).toBeDefined();
-    expect(res.headers["set-cookie"][0]).toContain("adminToken=;");
+    const logoutCookies = res.headers["set-cookie"];
+    expect(Array.isArray(logoutCookies) && logoutCookies[0]).toBeTruthy();
+    expect(logoutCookies![0]).toContain("adminToken=;");
   });
 });
 
@@ -88,11 +90,12 @@ describe("GET /api/auth/admin/check", () => {
       });
 
     const cookies = loginRes.headers["set-cookie"];
+    expect(Array.isArray(cookies) && cookies[0]).toBeTruthy();
 
     // Then check auth status
     const res = await request(app)
       .get("/api/auth/admin/check")
-      .set("Cookie", cookies)
+      .set("Cookie", cookies as string[])
       .expect(200);
 
     expect(res.body.authenticated).toBe(true);

--- a/apps/api/tsconfig.json
+++ b/apps/api/tsconfig.json
@@ -2,17 +2,19 @@
   "extends": "../../tsconfig.base.json",
   "compilerOptions": {
     "target": "ES2022",
-    "module": "NodeNext",
-    "moduleResolution": "NodeNext",
+    "module": "ESNext",
+    "moduleResolution": "node16",
+    "verbatimModuleSyntax": true,
+    "exactOptionalPropertyTypes": true,
     "strict": true,
     "skipLibCheck": true,
-
-    // Common flags
     "resolveJsonModule": true,
     "esModuleInterop": true,
     "allowSyntheticDefaultImports": true,
-    "verbatimModuleSyntax": true,
     "baseUrl": ".",
+    "paths": {
+      "@api/*": ["./src/*"]
+    },
     "types": ["node"]
   }
 }

--- a/apps/web/src/features/admin/components/Orders/OrdersTab.tsx
+++ b/apps/web/src/features/admin/components/Orders/OrdersTab.tsx
@@ -1,5 +1,6 @@
 // apps/web/src/components/admin/OrdersTab.tsx
-import { useState, useEffect, useCallback, useMemo, type ComponentType } from 'react';
+import { useState, useEffect, useCallback, useMemo } from 'react';
+import type { ComponentType } from 'react';
 import {
   Package,
   User,
@@ -180,7 +181,7 @@ const OrdersTab = () => {
 
   // Status badge with explicit icon component type
   const StatusBadge = ({ status }: { status: OrderStatus }) => {
-    const statusConfig: Record<OrderStatus, { color: string; icon: ComponentType<{ className?: string }> }> = {
+    const statusConfig: Record<OrderStatus, { color: string; icon: ComponentType<React.SVGProps<SVGSVGElement>> }> = {
       pending: { color: 'bg-yellow-50 text-yellow-900 border border-yellow-200', icon: Clock },
       confirmed: { color: 'bg-blue-50 text-blue-900 border border-blue-200', icon: CheckCircle },
       completed: { color: 'bg-green-50 text-green-900 border border-green-200', icon: CheckCircle },

--- a/apps/web/src/features/shop/components/ShopCart.tsx
+++ b/apps/web/src/features/shop/components/ShopCart.tsx
@@ -2,7 +2,7 @@
 import React, { useMemo, useCallback } from 'react';
 import { CartModal, MiniCart, Checkout } from '@/features/shop/components';
 import { useCart, useVariationSelection } from '@/features/hooks';
-import type { StorefrontCard, CardVariation } from '@/types';
+import type { StorefrontCard, CardVariation, CartItem } from '@/types';
 
 interface ShopCartProps {
   cards: StorefrontCard[];
@@ -36,12 +36,12 @@ export const ShopCart: React.FC<ShopCartProps> = ({
 
   // Derived state
   const cartTotal = useMemo(() =>
-    cart.items.reduce((sum, item) => sum + (item.price * item.quantity), 0),
+    cart.items.reduce((sum: number, item: CartItem) => sum + (item.price * item.quantity), 0),
     [cart]
   );
 
   const cartCount = useMemo(() =>
-    cart.items.reduce((sum, item) => sum + item.quantity, 0),
+    cart.items.reduce((sum: number, item: CartItem) => sum + item.quantity, 0),
     [cart]
   );
 

--- a/apps/web/src/lib/utils/csv.ts
+++ b/apps/web/src/lib/utils/csv.ts
@@ -40,7 +40,11 @@ export const arrayToCSV = (data: Record<string, unknown>[], headers: string[] | 
   }
 
   // Use provided headers or extract from first object
-  const csvHeaders = headers || Object.keys(data[0]);
+  const firstRow = data[0];
+  if (!firstRow) {
+    return '';
+  }
+  const csvHeaders = headers || Object.keys(firstRow);
 
   // Create header row
   const headerRow = csvHeaders.map(header => `"${header}"`).join(',');
@@ -193,11 +197,28 @@ export const validateInventoryCSV = (data: Record<string, unknown>[]): Validatio
 
   if (!Array.isArray(data) || data.length === 0) {
     errors.push('No valid data found in CSV');
-    return { valid: false, errors, warnings };
+    return {
+      totalRows: 0,
+      validRows: 0,
+      valid: false,
+      errors,
+      warnings
+    };
   }
 
   // Check headers
-  const headers = Object.keys(data[0]);
+  const firstRow = data[0];
+  if (!firstRow) {
+    errors.push('First row is empty or invalid');
+    return {
+      totalRows: 0,
+      validRows: 0,
+      valid: false,
+      errors,
+      warnings
+    };
+  }
+  const headers = Object.keys(firstRow);
   const missingRequired = requiredFields.filter(field => !headers.includes(field));
 
   if (missingRequired.length > 0) {
@@ -210,37 +231,42 @@ export const validateInventoryCSV = (data: Record<string, unknown>[]): Validatio
 
     // Check required fields
     requiredFields.forEach(field => {
-      if (!row[field] || String(row[field]).trim() === '') {
+      const fieldValue = row?.[field];
+      if (!fieldValue || String(fieldValue).trim() === '') {
         errors.push(`Row ${rowNumber}: Missing required field '${field}'`);
       }
     });
 
     // Validate price
-    if (row.price) {
-      const price = parseFloat(row.price);
+    const priceValue = row?.price;
+    if (priceValue) {
+      const price = parseFloat(String(priceValue));
       if (isNaN(price) || price < 0) {
-        errors.push(`Row ${rowNumber}: Invalid price '${row.price}'`);
+        errors.push(`Row ${rowNumber}: Invalid price '${priceValue}'`);
       } else if (price > 1000) {
-        warnings.push(`Row ${rowNumber}: High price detected '${row.price}'`);
+        warnings.push(`Row ${rowNumber}: High price detected '${priceValue}'`);
       }
     }
 
     // Validate stock quantity
-    if (row.stock_quantity) {
-      const stock = parseInt(row.stock_quantity, 10);
+    const stockValue = row?.stock_quantity;
+    if (stockValue) {
+      const stock = parseInt(String(stockValue), 10);
       if (isNaN(stock) || stock < 0) {
-        errors.push(`Row ${rowNumber}: Invalid stock quantity '${row.stock_quantity}'`);
+        errors.push(`Row ${rowNumber}: Invalid stock quantity '${stockValue}'`);
       }
     }
 
     // Validate quality
-    if (row.quality && !validQualities.includes(row.quality)) {
-      errors.push(`Row ${rowNumber}: Invalid quality '${row.quality}'. Must be one of: ${validQualities.join(', ')}`);
+    const qualityValue = row?.quality;
+    if (qualityValue && !validQualities.includes(String(qualityValue))) {
+      errors.push(`Row ${rowNumber}: Invalid quality '${qualityValue}'. Must be one of: ${validQualities.join(', ')}`);
     }
 
     // Validate foil type
-    if (row.foil_type && !validFoilTypes.includes(row.foil_type)) {
-      warnings.push(`Row ${rowNumber}: Unusual foil type '${row.foil_type}'. Expected: ${validFoilTypes.join(', ')}`);
+    const foilTypeValue = row?.foil_type;
+    if (foilTypeValue && !validFoilTypes.includes(String(foilTypeValue))) {
+      warnings.push(`Row ${rowNumber}: Unusual foil type '${foilTypeValue}'. Expected: ${validFoilTypes.join(', ')}`);
     }
 
     // Check for suspicious duplicates

--- a/apps/web/src/shared/layout/ErrorBoundary.tsx
+++ b/apps/web/src/shared/layout/ErrorBoundary.tsx
@@ -1,5 +1,6 @@
 // apps/web/src/components/ErrorBoundary.tsx
-import { Component, ErrorInfo, ReactNode } from "react";
+import { Component } from "react";
+import type { ErrorInfo, ReactNode } from "react";
 import { errorLogger } from "@/lib/utils/errorLogger";
 
 type Props = { fallback?: ReactNode; children: ReactNode };

--- a/apps/web/src/types/models/card.ts
+++ b/apps/web/src/types/models/card.ts
@@ -276,4 +276,5 @@ export function getUniqueFinishes(cards: Card[]): string[] {
 // EXPORT ALL
 // ============================================================================
 
-export default Card;
+// Note: No default export with verbatimModuleSyntax
+// Use named exports: import { Card } from './card';

--- a/apps/web/tsconfig.json
+++ b/apps/web/tsconfig.json
@@ -2,16 +2,20 @@
 {
   "extends": "../../tsconfig.base.json",
   "compilerOptions": {
+    "module": "ESNext",
+    "moduleResolution": "bundler",
+    "verbatimModuleSyntax": true,
+    "exactOptionalPropertyTypes": true,
     "noEmit": true,
-    "baseUrl": ".",                       
+    "baseUrl": ".",
     "paths": {
       "@/*": ["./src/*"],
       "@/features/*": ["./src/features/*"],
       "@/shared/*": ["./src/shared/*"],
       "@/lib/*": ["./src/lib/*"],
       "@/services/*": ["./src/services/*"],
-      "@/types": ["./src/types"]     
-    },
+      "@/types": ["./src/types"]
+    }
   },
   "include": ["src", "vite.config.ts"],
   "exclude": ["src/retired/**/*", "src/testUtils/**/*"]

--- a/tsconfig.base.json
+++ b/tsconfig.base.json
@@ -7,7 +7,8 @@
     "moduleResolution": "Bundler",
     "baseUrl": ".",
     "paths": {
-      "@shared/*": ["utils/shared/src/*"]
+      "@/*": ["apps/web/src/*"],
+      "@api/*": ["apps/api/src/*"]
     },
     "allowJs": true,
     "checkJs": false,

